### PR TITLE
Allow to keep form validation when using custom save buttons

### LINKED ISSUE #4295: No input validations when submit on toolbar custom actions
I have a problem after migrating to v3 on a custom action on the `Toolbar`.

At the validation of my form, I have to make two calls to my API.
 
Before the migration I used the `handleSubmit `props on a button to get the form data and make these 2 calls.
In V3 I can't use the `handleSubmit `of react-final-form, I'm using the `useFormState `to retrieve the data. 

The problem is that in v2 I had the validations on each input who have the `validate` prop but in v3 I can submit the form with empty inputs, I don't have the validations anymore.

here my Toolbar code in v2 : 

```
class Toolbar extends Component {
  render() {
    const {
      customSaveButton,
      handleSubmit,
      handleSubmitWithRedirect,
      onSubmit,
      record,
      resource,
      saveButton,
      translate,
    } = this.props;

    return (
      <Root>
        <StyledDiv>
          {customSaveButton && (
            <SaveButton
              id="SLM_saveButton"
              label={translate('ui_save')}
              onClick={handleSubmit(data => onSubmit(data))}
              submitOnEnter
            />
          )}
          {saveButton && (
            <SaveButton
              handleSubmitWithRedirect={handleSubmitWithRedirect}
              id="SLM_saveButton"
              label={isNilOrEmpty(record) ? translate('ui_save') : translate('ui_update')}
              resource={resource}
              submitOnEnter
            />
          )}
        </StyledDiv>
      </Root>
    );
  }
}

export default Toolbar;
```

In v3 : 
```
const Toolbar = ({
  customSaveButton,
  handleSubmitWithRedirect,
  onSubmit,
  record,
  resource,
  saveButton,
  saving,
  translate,
  invalid,
}) => {
  const formState = useFormState();

  const handleCustomSubmit = useCallback(() => {
    onSubmit(formState.values, invalid);
  }, [formState.values]);

  return (
    <Root>
      <StyledDiv>
        {customSaveButton && (
          <Button id="SLM_saveButton" onClick={handleCustomSubmit}>
            <Save />
            {translate(customSaveButton ? 'ui_save' : 'ui_update')}
          </Button>
        )}
        {saveButton && (
          <SaveButton
            handleSubmitWithRedirect={handleSubmitWithRedirect}
            id="SLM_saveButton"
            label={translate(isNilOrEmpty(record) ? 'ui_save' : 'ui_update')}
            resource={resource}
            saving={saving}
            submitOnEnter
            variant="text"
          />
        )}
      </StyledDiv>
    </Root>
  );
};
```

And here the component who call api, the `CreateBaseCampaign ` contain inputs with `validate `prop :

```
class CampaignCreate extends Component {
  onSubmit = (data, invalid) => {
    const {
      createCampaign,
      createLineItem,
      history: { push },
    } = this.props;
    const { levers } = data;

    if (!invalid) {
      createCampaign(
        `${CAMPAIGNS}`,
        {
          ...data,
        },
        {},
        false,
        ({
          payload: {
            data: { id },
          },
        }) =>
          createLineItem(
            `${CAMPAIGNS}/${id}/lineitems`,
            levers,
            {},
            { text: 'campaign_created', level: 'info' },
            () => push(`/${CAMPAIGNS}/${id}`),
          ),
      );
    }
  };

  render() {
    return (
      <Create {...this.props}>
        <SimpleForm
          toolbar={
            <Toolbar
              url={`/${CAMPAIGNS}`}
              customSaveButton
              saveButton={false}
              onSubmit={this.onSubmit}
            />
          }
          redirect={null}
        >
          <CreateBaseCampaign {...this.props} />
        </SimpleForm>
      </Create>
    );
   }
 }

 export default CampaignCreate;
```

**Environment**

* React-admin version: v3.1.2
* Last version that did not exhibit the issue (if applicable): 2.8
* React version: 16.12.0
* Browser:
* Stack trace (in case of a JS error):

### PR DESCRIPTION
**Pull request by JulienMattiussi**
Closes #4295 

## Issue

Form validation is never performed when using custom save buttons
(doing custom dava provider instead of dfault ra REST call)

Because when we overload handleSubmit or handleSubmitWithRedirect function, the form submission isn't performed

(And react-final-form doesn't offer a simple possibility to perform a form validation without a form submission.)

## Solution

Add possibility to overload ra save function without overloading handleSubmit function

Use this
```js
const SaveCustomButton = props => {
    const [create] = useCreate('posts');
    const redirectTo = useRedirect();
    const notify = useNotify();
    const { basePath } = props;

    const handleSave = useCallback(
        (values, redirect) => {
            create(
                {
                    payload: {
                        data: { ...values, average_note: 10 },
                    },
                },
                {
                    onSuccess: ({ data: newRecord }) => {
                        notify('ra.notification.created', 'info', {
                            smart_count: 1,
                        });
                        redirectTo(redirect, basePath, newRecord.id, newRecord);
                    },
                }
            );
        },
        [create, notify, redirectTo, basePath]
    );

    return <SaveButton {...props} onSave={handleSave} />;
};
```

Instead of this

```js
const SaveCustomButton = props => {
    const [create] = useCreate('posts');
    const redirectTo = useRedirect();
    const notify = useNotify();
    const formState = useFormState();
    const { basePath, redirect } = props;

    const handleClick = useCallback(
        () => {
            create(
                {
                    payload: {
                        data: { ...formState.values, average_note: 10 },
                    },
                },
                {
                    onSuccess: ({ data: newRecord }) => {
                        notify('ra.notification.created', 'info', {
                            smart_count: 1,
                        });
                        redirectTo(redirect, basePath, newRecord.id, newRecord);
                    },
                }
            );
        },
        [formState, redirect, create, notify, redirectTo, basePath]
    );

    return <SaveButton {...props} handleSubmitWithRedirect={handleClick} />;
};
```

## Doc

image

image

### CODE REVIEW

File: examples/simple/src/posts/PostCreate.js [hunk 1]
```diff
@@ -20,46 +20,36 @@ import {
...
+                        redirectTo(redirect, basePath, newRecord.id, newRecord);
+                    },
+                }
+            );
+        },
+        [create, notify, redirectTo, basePath]
+    );
 
-    return <SaveButton {...props} handleSubmitWithRedirect={handleClick} />;
+    return <SaveButton {...props} customSave={handleClick} />;
```
Comment (fzaninotto) [inline examples/simple/src/posts/PostCreate.js]: let's rename it to `onSave`

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 2]
```diff
@@ -49,13 +50,18 @@ const FormWithRedirect = ({
...
     let redirect = useRef(props.redirect);
+    let realSave = useRef(save);
     // We don't use state here for two reasons:
     // 1. There no way to execute code only after the state has been updated
     // 2. We don't want the form to rerender when redirect is changed
     const setRedirect = newRedirect => {
         redirect.current = newRedirect;
     };
 
+    const setSave = newSave => {
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx]: setOnSave

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 3]
```diff
@@ -49,13 +50,18 @@ const FormWithRedirect = ({
     ...props
 }) => {
     let redirect = useRef(props.redirect);
+    let realSave = useRef(save);
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx]: let onSave

File: yarn.lock [hunk 4]
```diff
@@ -1579,7 +1579,7 @@
   resolved "https://registry.yarnpkg.com/@types/parse-json/-/parse-json-4.0.0.tgz#2f8bb441434d163b35fb8ffdccd7138927ffb8c0"
```
[the remark this answers is not in the document]
Comment (JulienMattiussi) [reply yarn.lock]: Maybe internal dependencies when I switched from master to next.
I can revert it

[the remark this answers is not in the document]
Comment (fzaninotto) [reply yarn.lock]: yes, please

[the remark this answers is not in the document]
Comment (JulienMattiussi) [reply yarn.lock]: done

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 5]
```diff
@@ -38,13 +38,13 @@ const SaveButton: FC<SaveButtonProps> = ({
     const classes = useStyles({ classes: classesOverride });
     const notify = useNotify();
     const translate = useTranslate();
-    const setOnSave = useContext(OnSaveContext);
+    const formContext = useContext(FormContext);
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/button/SaveButton.tsx]:
```suggestion
    const { setOnSave } = useContext(FormContext);
```

File: packages/ra-core/src/form/FormWithRedirect.tsx [hunk 6]
```diff
@@ -83,7 +83,7 @@ const FormWithRedirect = ({
     };
 
     return (
-        <OnSaveContext.Provider value={setOnSave}>
+        <FormContext.Provider value={{ setOnSave }}>
```
Comment (fzaninotto) [inline packages/ra-core/src/form/FormWithRedirect.tsx]: Add memoization:

const formContextValue = useMemo(() => ({ setOnSave }), [setOnSave]);

**Pull request merged**